### ISSUE TITLE
async within sync call warning and override

### ISSUE THREAD

**Issue by bsz0206** [dart-lang/linter#4787]
when an async method is called from a sync function it should warn and the developer should specify it is an informed decision:

Bad:
class B {
   vod doTheJob() async { ... }
}
void f() {
  B.get();
}

Good:
class B {
   vod doTheJob() async { ... }
}
void f() {
  awaited(B.get());
}

**Comment by srawlins** [comment]
I believe this is the lint rule you are looking for: https://dart.dev/tools/linter-rules/discarded_futures

**Comment by lrhn** [comment]
Actually, as described, the `discarded_futures` lint makes you turn the `f` function into an `async` one.
It doesn't give you a way to say that the call is deliberate.

**Comment by srawlins** [comment]
Ah, that would be a comment like `// ignore: discared_futures`.

**Comment by bsz0206** [comment]
> Ah, that would be a comment like `// ignore: discared_futures`.

Such comments are for exceptional or temporary basis. This is a real case and needs support.

**Comment by bsz0206** [comment]
unawaid currently is a "void unwaited(...) and needs to turn into a "dynamic unwaited(...)" to provide support.

**Comment by lrhn** [comment]
There is no lint which does for `sync` functions what `unawaited_futures` does for `async` functions.
It would probably be nice to have one.

The `discarded_futures` lint is something else. It's misnamed, but it really is "no calling async functions from a `sync` function" ... at all! The remedies are converting the `sync` function to an `async` function (which enables `unawaited_futures`), not calling the `async` function at all, or using `// ignore`. (According to the documentation, which is as close to a specification as we have.)

I agree that `// ignore:` should only be for exceptional cases. If you have many instances, then the lint is probably not a good match for your code. (Which is likely for `discared_futures`.)

About the original example:
```drt
class B {
  void doTheJob() async { ... }
}
```
This `doTheJob` method returns `void`, so it's unlikely and undesirable that the calling method should consider the method asynchronous. It's *written* to make everybody ignore the returned future, and the linter should consider that a good enough reason to actually ignore it.


There are no plans to change `unawaited`. It's designed for `unawaited_futures`, and makes it so that there is no expression evaluating to a `Future`, which isn't also used in some way. If it returned a `Future`, it wouldn't work for that purpose. (Returning `dynamic` is right out. People don't want `dynamic`.)

**Comment by tlserver** [comment]
> There is no lint which does for sync functions what unawaited_futures does for async functions.
> It would probably be nice to have one.

Yes, I agree that it is nice to have one lint which does for sync functions what unawaited_futures does for async functions and `discarded_futures` is misnamed.

I think it is too strict for rule "no calling async functions from a sync function". Since a sync function can have a parameter having `Future` type, such as `unawaited()` and `FutureBuilder()` in flutter. Is there any usecases that need this strict rule? Maybe most users do not want to be warned while calling `unawaited(asyncFunc)` or `FutureBuilder(future: asyncFunc, ...)` in a sync function. Please note that calling `unawaited(asyncFunc)` in a sync function is not warned in current behavior but it does break the description "no calling async functions from a sync function".

Beside, the [document of `unawaited_futures`](https://dart.dev/tools/linter-rules/unawaited_futures) is also not match the current behavior. It state:
> Future results in async function bodies must be awaited or marked unawaited using dart:async.

Base on this description, 
```dart
  void a(Future future) {}
  Future<void> b() async {}
  Future<void> c() async {
    a(b());
  }
```
should be warned, since `b()` is neither awaited nor marked `unawaited()`.

This behavior is nice and should be kept. But to better descripted this behavior, maybe "Future results in async blocks must be used"? For example:
```dart
  void a(Future future) {}
  Future<void> b() async {}
  Future<void> c() async {
    a(b()); // Future instance b() is used as a argument of a(), so no warning
    unawaited(b()); // used as a argument of unawaited()
    await b(); // used by await keyword
    b().then(() {}).ignore() // b() and then() are both used. ignore() is not a future
    
    b() // **warning**, b() is not used
    b().then(() {}) // **warning**, b() is used by then(), but then() is not used
    Future(() {}) // **warning**, not an async function, but a future factory constructor. So the new description match better
  }
```

`unawaited_futures` is good, but it only warn for async blocks. Most people may want a lint that defined as "Future results *~in async blocks~* must be used". And I think this rule match the name `discarded_futures`.

But then, this rule is overlapped with `unawaited_futures`. So one possible solution is changing the description to "Future results in sync blocks must be used". Or just extend the `unawaited_futures` to check for both sync and async blocks. It is rare that someone need to check unused future only for async blocks but not sync blocks.

Just my opinion. if I am wrong, feel free to correct me.
